### PR TITLE
Interfaces in ADO.Net

### PR DESCRIPTION
**Pull request by saurabh500**
This PR aims at adding the following interfaces in ADO.Net and expose them in contracts .
IDataParameter 
IDataParameterCollection 
IDataReader
IDataRecord 
IDbCommand 
IDbConnection 
IDbDataParameter 
IDbTransaction

### CODE REVIEW

**Comment by saurabh500** [comment]
@kkurni @YoungGah @corivera @nadiatk

File: src/System.Data.Common/ref/System.Data.Common.cs [hunk 1]
```diff
@@ -45,6 +45,17 @@ public enum ConnectionState
...
+    public enum DataRowVersion
+    {
+        Current = 512,
+        Default = 1536,
+        Original = 256,
+        Proposed = 1024,
+    }
+    public partial class DataTable
+    {
+        internal DataTable() { }
```
Comment (kkurni) [inline src/System.Data.Common/ref/System.Data.Common.cs]: Just wondering why do we need internal constructor for DataTable ?

Comment (saurabh500) [inline src/System.Data.Common/ref/System.Data.Common.cs]: This was auto-generated by the Contract Generation tool.

**Comment by saurabh500** [comment]
cc @terrajobst @ericstj

File: src/System.Data.Common/src/System/Data/DataTable.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// ------------------------------------------------------------------------------
+
+namespace System.Data {
+   
+    // An empty Datatable class to allow complete interfaces
+    public class DataTable{
+
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/DataTable.cs]: Is it intentional to have an internal constructor in the reference assembly but a public constructor here in the implementation? (By not specifying a constructor here, the C# compiler will generate a public default constructor).

File: src/System.Data.Common/src/System/Data/Common/DbDataRecord.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+
+//------------------------------------------------------------------------------
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/Common/DbDataRecord.cs]: Nit: remove this comment and unnecessary lines?

File: src/System.Data.Common/src/System/Data/DataRowVersion.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+namespace System.Data
+{
+    public enum DataRowVersion {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/DataRowVersion.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/DataTable.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// ------------------------------------------------------------------------------
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/DataTable.cs]: Nit: remove the "// -----" comment

File: src/System.Data.Common/src/System/Data/DataTable.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace System.Data {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/DataTable.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDataParameter.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+namespace System.Data {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDataParameter.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDataParameterCollection.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+
+namespace System.Data {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDataParameterCollection.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDataReader.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
+//------------------------------------------------------------------------------
+// <copyright file="IDataReader.cs" company="Microsoft">
+//      Copyright (c) Microsoft Corporation.  All rights reserved.
+// </copyright>
+// <owner current="true" primary="true">markash</owner>
+// <owner current="true" primary="false">laled</owner>
+//------------------------------------------------------------------------------
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDataReader.cs]: Nit: incorrect license header

File: src/System.Data.Common/src/System/Data/IDataReader.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace System.Data {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDataReader.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDbCommand.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+namespace System.Data{
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDbCommand.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDbConnection.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+
+namespace System.Data {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDbConnection.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDbDataParameter.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+namespace System.Data {
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDbDataParameter.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

File: src/System.Data.Common/src/System/Data/IDbTransaction.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
+namespace System.Data{
```
Comment (justinvp) [inline src/System.Data.Common/src/System/Data/IDbTransaction.cs]: Nit: the coding style for corefx is Allman style braces, where the braces start on a new line.

Comment (saurabh500) [inline src/System.Data.Common/src/System/Data/IDbTransaction.cs]: I will change it. I had copied over the code from .Net 4.5. 
Thanks for the feedback

**Comment by kkurni** [comment]
LGTM

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by saurabh500** [comment]
@dotnet-bot test this please

**Comment by terrajobst** [comment]
LGTM

**Comment by saurabh500** [comment]
https://github.com/dotnet/corefx/issues/3480#issuecomment-188577701

**Comment by nvivo** [comment]
One thing that is not making sense to me here is the lack of async methods in these interfaces. 

Since async methods *should* be the main way to  acess a database nowadays,  not exposing async methods them would make the interfaces useless. They'd be just a compatibility thing. 

Also,  it would be much better exposing something more complete like ExecuteReader(CommandBehavior) and ExecuteReaderAsync(CommandBehavior,  CancellationToken)  than all overloads for the sync/async variations. If overloads are added,  they should either be added for sync and async versions,  or for none.

 I personally think interfaces should be as dry as possible. Only exposing the full signatures for the methods that matter would be cleaner and more useful for anyone using them.

**Comment by CurtHagenlocher** [comment]
Does the reference assembly need to be updated too? I see that line 301 of System.Data.Common.cs still doesn't include the interface as a base.

**Pull request merged**